internal/dnf/dnf.go

```diff
@@ -0,0 +1,262 @@
+	{Path: `usr/lib/sysimage/libdnf5/transaction_history.sqlite`, Enum: 5}, // dnf5
+	// Defined here:
+	// https://github.com/rpm-software-management/libdnf/blob/93759bc5cac262906e52b6a173d7b157914ec29e/libdnf/transaction/Types.hpp#L45
+	{Path: `var/lib/dnf/history.sqlite`, Enum: 8}, // dnf3/4
+}
+
+// HistoryDB is a handle to the dnf history database.
+//
+// All methods are safe to call on a nil receiver.
+type historyDB struct {
+	db *sql.DB
+	rm int
+}
+
+// OpenHistoryDB does what it says on the tin.
+//
+// This function may return a nil *historyDB, which is still safe to use.
+func openHistoryDB(ctx context.Context, sys fs.FS) (*historyDB, error) {
+	var found *dbDesc
+	for _, v := range possible {
+		switch _, err := fs.Stat(sys, v.Path); {
+		case errors.Is(err, nil):
+			found = &v
+		case errors.Is(err, fs.ErrNotExist): // OK
+		default:
+			return nil, fmt.Errorf("internal/dnf: unexpected error handling fs.FS: %w", err)
+		}
+	}
+	if found == nil {
+		return nil, nil
+	}
+	var h *historyDB
+
+	f, err := sys.Open(found.Path)
+	if err != nil {
+		return nil, fmt.Errorf("internal/dnf: unexpected error opening history: %w", err)
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			zlog.Warn(ctx).Err(err).Msg("unable to close fs.FS db file")
+		}
+	}()
+
+	// Needs to be linked into the filesystem for the sqlite driver to open it.
+	// See also: quay/claircore#720
+	spool, err := os.CreateTemp(os.TempDir(), `dnf.sqlite.*`)
+	if err != nil {
+		return nil, fmt.Errorf("internal/dnf: error reading sqlite db: %w", err)
+	}
+	defer func() {
+		if err := spool.Close(); err != nil {
+			zlog.Warn(ctx).Err(err).Msg("unable to close sqlite db file")
+		}
+		// If in an error return, make sure to clean up the spool file.
+		if h == nil {
+			if err := os.Remove(spool.Name()); err != nil {
+				zlog.Warn(ctx).Err(err).Msg("unable to unlink sqlite db file")
+			}
+		}
+	}()
+
+	zlog.Debug(ctx).Str("file", spool.Name()).Msg("copying sqlite db out of tar")
+	if _, err := io.Copy(spool, f); err != nil {
+		return nil, fmt.Errorf("internal/dnf: error spooling sqlite db: %w", err)
+	}
+	if err := spool.Sync(); err != nil {
+		return nil, fmt.Errorf("internal/dnf: error spooling sqlite db: %w", err)
+	}
+
+	db, err := sql.Open("sqlite", spool.Name())
+	if err != nil {
+		return nil, fmt.Errorf("internal/dnf: error reading sqlite db: %w", err)
+	}
+	if err := db.PingContext(ctx); err != nil {
+		return nil, fmt.Errorf("internal/dnf: error reading sqlite db: %w", err)
+	}
+
+	h = &historyDB{
+		db: db,
+		rm: found.Enum,
+	}
+	runtime.AddCleanup(db, func(p string) {
+		os.Remove(p)
+	}, spool.Name())
+	// This is an internal function, so add an extra caller frame.
+	_, file, line, _ := runtime.Caller(2)
+	runtime.SetFinalizer(h, func(_ *historyDB) {
+		panic(fmt.Sprintf("%s:%d: historyDB not closed", file, line))
```

Comment:
idea: can we just `Close` for those who may forget?

Reply:
No, the point of this is to help callers not forget.